### PR TITLE
config: fix the store limit cannot be persisted

### LINKED ISSUE #3382: Store limit isn't persist when we restart the server
### What did you do?

1. start a cluster with 1 PD, 1TiKV
2. change the store limit to 30 by using pd-ctl
3. restart the PD server
4. show store limit

### What did you expect to see?

The store limit is changed to 30.

### What did you see instead?

The result is 15 instead.

### What version of PD are you using (`pd-server -V`)?
since v4.0.2

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

The store limit cannot persist due to we only change the value in the memory. Close #3382.

### What is changed and how it works?

This PR persists the store limit once it changes and if persistence fails, we will roll back to the original one.

### Check List

Tests

- Unit test

Related changes

- Need to cherry-pick to the release branch

### Release note

- Fix the issue that the store limit cannot be persisted

### CODE REVIEW

File: server/cluster/cluster.go [hunk 1]
```diff
@@ -1651,16 +1654,35 @@ func (c *RaftCluster) RemoveStoreLimit(storeID uint64) {
...
+	}
 }
 
 // SetStoreLimit sets a store limit for a given type and rate.
 func (c *RaftCluster) SetStoreLimit(storeID uint64, typ storelimit.Type, ratePerMin float64) {
+	old := c.opt.GetScheduleConfig().Clone()
 	c.opt.SetStoreLimit(storeID, typ, ratePerMin)
+	if err := c.opt.Persist(c.storage); err != nil {
+		// roll back the store limit
+		c.opt.SetScheduleConfig(old)
```
Comment (JmPotato) [inline server/cluster/cluster.go:1668]: What will happen if the rollback operation fails also?

Comment (rleungx) [reply server/cluster/cluster.go:1668]: The rollback operation only resets the memory value

**Review by Yisaer — APPROVED** [review]
LGTM

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -1651,16 +1659,39 @@ func (c *RaftCluster) RemoveStoreLimit(storeID uint64) {
...
+		log.Error("persist store limit meet error", errs.ZapError(err))
+	}
 }
 
 // SetAllStoresLimit sets all store limit for a given type and rate.
 func (c *RaftCluster) SetAllStoresLimit(typ storelimit.Type, ratePerMin float64) {
```
Comment (HunDunDM) [inline server/cluster/cluster.go:1683]: ditto

File: server/cluster/cluster.go [hunk 3]
```diff
@@ -1641,6 +1642,13 @@ func (c *RaftCluster) AddStoreLimit(store *metapb.Store) {
 
 	cfg.StoreLimit[storeID] = sc
 	c.opt.SetScheduleConfig(cfg)
+	var err error
+	for i := 0; i < persistLimitRetryTime; i++ {
```
Comment (HunDunDM) [inline server/cluster/cluster.go:1646]: Add some comment and TODO?

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -1651,16 +1659,39 @@ func (c *RaftCluster) RemoveStoreLimit(storeID uint64) {
 	}
 	delete(cfg.StoreLimit, storeID)
 	c.opt.SetScheduleConfig(cfg)
+	var err error
+	for i := 0; i < persistLimitRetryTime; i++ {
```
Comment (HunDunDM) [inline server/cluster/cluster.go:1663]: ditto

File: server/cluster/cluster.go [hunk 5]
```diff
@@ -1651,16 +1659,39 @@ func (c *RaftCluster) RemoveStoreLimit(storeID uint64) {
...
+		if err = c.opt.Persist(c.storage); err == nil {
+			return
+		}
+	}
+	log.Error("persist store limit meet error", errs.ZapError(err))
 }
 
 // SetStoreLimit sets a store limit for a given type and rate.
 func (c *RaftCluster) SetStoreLimit(storeID uint64, typ storelimit.Type, ratePerMin float64) {
```
Comment (HunDunDM) [inline server/cluster/cluster.go:1672]: Need to return an error and handle it?

**Comment by rleungx** [comment]
I made some new changes, PTAL @Yisaer @JmPotato

**Review by HunDunDM — APPROVED** [review]

**Review by Yisaer — APPROVED** [review]
LGTM

File: server/cluster/cluster.go [hunk 6]
```diff
@@ -56,6 +56,10 @@ var backgroundJobInterval = 10 * time.Second
 const (
 	clientTimeout              = 3 * time.Second
 	defaultChangedRegionsLimit = 10000
+	// persistLimitRetryTime is used to reduce the probability of the persistent error
+	// since the once the store is add or remove, we shouldn't return an error even if the store limit is failed to persist.
+	persistLimitRetryTime = 5
```
Comment (nolouch) [inline server/cluster/cluster.go:61]:
```suggestion
	persistLimitRetryTimes = 5
```

**Pull request merged**